ethcore/src/client/traits.rs

```diff
@@ -46,80 +49,198 @@ use types::block_status::BlockStatus;
+		StateOrBlock::State(info)
+	}
+}
+
+impl From<BlockId> for StateOrBlock {
+	fn from(id: BlockId) -> StateOrBlock {
+		StateOrBlock::Block(id)
+	}
+}
+
+/// Provides `nonce` and `latest_nonce` methods
+pub trait Nonce {
+	/// Attempt to get address nonce at given block.
+	/// May not fail on BlockId::Latest.
+	fn nonce(&self, address: &Address, id: BlockId) -> Option<U256>;
 
+	/// Get address nonce at the latest block's state.
+	fn latest_nonce(&self, address: &Address) -> U256 {
+		self.nonce(address, BlockId::Latest)
+			.expect("nonce will return Some when given BlockId::Latest. nonce was given BlockId::Latest. \
+			Therefore nonce has returned Some; qed")
+	}
+}
+
+/// Provides `balance` and `latest_balance` methods
+pub trait Balance {
+	/// Get address balance at the given block's state.
+	///
+	/// May not return None if given BlockId::Latest.
+	/// Returns None if and only if the block's root hash has been pruned from the DB.
+	fn balance(&self, address: &Address, state: StateOrBlock) -> Option<U256>;
+
+	/// Get address balance at the latest block's state.
+	fn latest_balance(&self, address: &Address) -> U256 {
+		self.balance(address, BlockId::Latest.into())
+			.expect("balance will return Some if given BlockId::Latest. balance was given BlockId::Latest \
+			Therefore balance has returned Some; qed")
+	}
+}
+
+/// Provides methods to access account info
+pub trait AccountData: Nonce + Balance {}
+
+/// Provides `chain_info` method
+pub trait ChainInfo {
+	/// Get blockchain information.
+	fn chain_info(&self) -> BlockChainInfo;
+}
+
+/// Provides various information on a block by it's ID
+pub trait BlockInfo {
 	/// Get raw block header data by block id.
 	fn block_header(&self, id: BlockId) -> Option<encoded::Header>;
 
+	/// Get the best block header.
+	fn best_block_header(&self) -> encoded::Header;
+
+	/// Get raw block data by block header hash.
+	fn block(&self, id: BlockId) -> Option<encoded::Block>;
+
+	/// Get address code hash at given block's state.
+	fn code_hash(&self, address: &Address, id: BlockId) -> Option<H256>;
+}
+
+/// Provides various information on a transaction by it's ID
+pub trait TransactionInfo {
+	/// Get the hash of block that contains the transaction, if any.
+	fn transaction_block(&self, id: TransactionId) -> Option<H256>;
+}
+
+/// Provides methods to access chain state
+pub trait StateClient {
+	/// Type representing chain state
+	type State: StateInfo;
+
+	/// Get a copy of the best block's state.
+	fn latest_state(&self) -> Self::State;
+
+	/// Attempt to get a copy of a specific block's final state.
+	///
+	/// This will not fail if given BlockId::Latest.
+	/// Otherwise, this can fail (but may not) if the DB prunes state or the block
+	/// is unknown.
+	fn state_at(&self, id: BlockId) -> Option<Self::State>;
+}
+
+/// Provides various blockchain information, like block header, chain state etc.
+pub trait BlockChain: ChainInfo + BlockInfo + TransactionInfo /* + StateClient */ {}
```

Comment:
Is the `StateClient` bound desired? Why is it commented?

Reply:
Nope, it's an artifact. Previously I thought, that `BlockChain` should bound on `StateClient`  but it wasn't the case.